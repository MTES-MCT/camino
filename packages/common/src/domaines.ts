export const DOMAINES_IDS = [
  "c",
  "f",
  "g",
  "h",
  "i",
  "m",
  "r",
  "s",
  "w",
] as const;

export type DomaineId = typeof DOMAINES_IDS[number];

interface Domaine {
  id: DomaineId;
  nom: string;
  description: string;
  ordre: number;
}

export const Domaines: { [key in DomaineId]: Omit<Domaine, "id"> } = {
  c: {
    nom: "carrières",
    description:
      "Domaine couvrant les zones spéciales de carrières à terre qui incluent des substances minérales ou fossiles de carrières rares ou nécessaires pour satisfaire les besoins des consommateurs et l’intérêt économique national ou régional.",
    ordre: 3,
  },
  f: {
    nom: "combustibles fossiles",
    description:
      "Domaine minier auquel appartiennent les combustibles fossiles, la tourbe exceptée.",
    ordre: 5,
  },
  g: {
    nom: "géothermie",
    description:
      "Domaine auquel appartiennent les gîtes renfermés dans le sein de la terre dont on peut extraire de l'énergie sous forme thermique, notamment par l'intermédiaire des eaux chaudes et des vapeurs souterraines qu'ils contiennent.",
    ordre: 7,
  },
  h: {
    nom: "hydrocarbures liquides ou gazeux",
    description:
      "Domaine minier auquel appartiennent des hydrocarbures qu'ils soient sous forme liquide ou gazeuse.",
    ordre: 4,
  },
  i: {
    nom: "indéterminé",
    description: "Domaine indéterminé",
    ordre: 9,
  },
  m: {
    nom: "minéraux et métaux",
    description:
      "Domaine minier auquel appartiennent les substances minérales non énergétiques et non radioactives classées dans la catégorie des substances de mines, notemment certains métaux, gaz, sels et terres-rares.",
    ordre: 1,
  },
  r: {
    nom: "éléments radioactifs",
    description:
      "Domaine minier auquel appartiennent les substances minérales radioactives classés dans la catégorie des substances de mines.",
    ordre: 6,
  },
  s: {
    nom: "stockages souterrains",
    description:
      "Domaine auquel appartiennent les cavités souterraines naturelles ou artificielles ou formations souterraines naturelles présentant les qualités requises pour constituer des réservoirs étanches ou susceptibles d'être rendus tels, en vue du stockage de gaz naturel, d'hydrocarbures liquides, liquéfiés ou gazeux ou de produits chimiques à destination industrielle.",
    ordre: 8,
  },
  w: {
    nom: "granulats marins",
    description:
      "Domaine auquel appartiennent les substances de carrières situés dans les fonds marins du domaine public.",
    ordre: 2,
  },
};
