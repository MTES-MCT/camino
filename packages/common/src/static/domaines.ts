import { Definition } from '../definition.js'

export const DOMAINES_IDS = {
  METAUX: 'm',
  CARRIERES: 'c',
  FOSSILES: 'f',
  GEOTHERMIE: 'g',
  HYDROCARBURE: 'h',
  RADIOACTIF: 'r',
  SOUTERRAIN: 's',
  GRANULATS_MARINS: 'w',
  INDETERMINE: 'i'
} as const

export type DomaineId = typeof DOMAINES_IDS[keyof typeof DOMAINES_IDS]
export type Domaine<T = DomaineId> = Definition<T>
export const Domaines: { [key in DomaineId]: Domaine<key> } = {
  c: {
    id: 'c',
    nom: 'carrières',
    description:
      'Domaine couvrant les zones spéciales de carrières à terre qui incluent des substances minérales ou fossiles de carrières rares ou nécessaires pour satisfaire les besoins des consommateurs et l’intérêt économique national ou régional.',
    ordre: 3
  },
  f: {
    id: 'f',
    nom: 'combustibles fossiles',
    description: 'Domaine minier auquel appartiennent les combustibles fossiles, la tourbe exceptée.',
    ordre: 5
  },
  g: {
    id: 'g',
    nom: 'géothermie',
    description:
      "Domaine auquel appartiennent les gîtes renfermés dans le sein de la terre dont on peut extraire de l'énergie sous forme thermique, notamment par l'intermédiaire des eaux chaudes et des vapeurs souterraines qu'ils contiennent.",
    ordre: 7
  },
  h: {
    id: 'h',
    nom: 'hydrocarbures liquides ou gazeux',
    description: "Domaine minier auquel appartiennent des hydrocarbures qu'ils soient sous forme liquide ou gazeuse.",
    ordre: 4
  },
  i: {
    id: 'i',
    nom: 'indéterminé',
    description: 'Domaine indéterminé',
    ordre: 9
  },
  m: {
    id: 'm',
    nom: 'minéraux et métaux',
    description:
      'Domaine minier auquel appartiennent les substances minérales non énergétiques et non radioactives classées dans la catégorie des substances de mines, notemment certains métaux, gaz, sels et terres-rares.',
    ordre: 1
  },
  r: {
    id: 'r',
    nom: 'éléments radioactifs',
    description: 'Domaine minier auquel appartiennent les substances minérales radioactives classés dans la catégorie des substances de mines.',
    ordre: 6
  },
  s: {
    id: 's',
    nom: 'stockages souterrains',
    description:
      "Domaine auquel appartiennent les cavités souterraines naturelles ou artificielles ou formations souterraines naturelles présentant les qualités requises pour constituer des réservoirs étanches ou susceptibles d'être rendus tels, en vue du stockage de gaz naturel, d'hydrocarbures liquides, liquéfiés ou gazeux ou de produits chimiques à destination industrielle.",
    ordre: 8
  },
  w: {
    id: 'w',
    nom: 'granulats marins',
    description: 'Domaine auquel appartiennent les substances de carrières situés dans les fonds marins du domaine public.',
    ordre: 2
  }
}

export const sortedDomaines = Object.values(Domaines).sort((a, b) => a.ordre - b.ordre)

export const domainesIds = Object.values(DOMAINES_IDS)

export const isDomaineId = (domaineId: string | null | undefined): domaineId is DomaineId => domainesIds.includes(domaineId)
