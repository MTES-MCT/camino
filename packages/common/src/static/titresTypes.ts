import { DomaineId, isDomaineId } from './domaines'
import { isTitreTypeType, TitreTypeTypeId } from './titresTypesTypes'
import { z } from 'zod'
interface Definition<T> {
  id: T
  domaineId: DomaineId
  typeId: TitreTypeTypeId
}
// prettier-ignore
const TITRES_TYPES_KEYS = ['apc', 'aph', 'apm', 'apw', 'arc', 'arg', 'arm', 'axm', 'cxf', 'cxg', 'cxh', 'cxm', 'cxr', 'cxs', 'cxw', 'inm', 'inr', 'pcc', 'prf', 'prg', 'prh', 'prm', 'prr', 'prs', 'prw', 'pxf', 'pxg', 'pxh', 'pxm', 'pxr', 'pxw',] as const

export const TITRES_TYPES_IDS = {
  AUTORISATION_DE_PROSPECTION_CARRIERES: 'apc',
  AUTORISATION_DE_PROSPECTION_HYDROCARBURE: 'aph',
  AUTORISATION_DE_PROSPECTION_METAUX: 'apm',
  AUTORISATION_DE_PROSPECTION_GRANULATS_MARINS: 'apw',

  AUTORISATION_DE_RECHERCHE_CARRIERES: 'arc',
  AUTORISATION_DE_RECHERCHE_GEOTHERMIE: 'arg',
  AUTORISATION_DE_RECHERCHE_METAUX: 'arm',

  AUTORISATION_D_EXPLOITATION_METAUX: 'axm',

  CONCESSION_FOSSILES: 'cxf',
  CONCESSION_GEOTHERMIE: 'cxg',
  CONCESSION_HYDROCARBURE: 'cxh',
  CONCESSION_METAUX: 'cxm',
  CONCESSION_RADIOACTIF: 'cxr',
  CONCESSION_SOUTERRAIN: 'cxs',
  CONCESSION_GRANULATS_MARINS: 'cxw',
  INDETERMINE_METAUX: 'inm',
  INDETERMINE_RADIOACTIF: 'inr',
  PERMIS_EXCLUSIF_DE_CARRIERES_CARRIERES: 'pcc',

  PERMIS_EXCLUSIF_DE_RECHERCHES_FOSSILES: 'prf',
  PERMIS_EXCLUSIF_DE_RECHERCHES_GEOTHERMIE: 'prg',
  PERMIS_EXCLUSIF_DE_RECHERCHES_HYDROCARBURE: 'prh',
  PERMIS_EXCLUSIF_DE_RECHERCHES_METAUX: 'prm',
  PERMIS_EXCLUSIF_DE_RECHERCHES_RADIOACTIF: 'prr',
  PERMIS_EXCLUSIF_DE_RECHERCHES_SOUTERRAIN: 'prs',
  PERMIS_EXCLUSIF_DE_RECHERCHES_GRANULATS_MARINS: 'prw',

  PERMIS_D_EXPLOITATION_FOSSILES: 'pxf',
  PERMIS_D_EXPLOITATION_GEOTHERMIE: 'pxg',
  PERMIS_D_EXPLOITATION_HYDROCARBURE: 'pxh',
  PERMIS_D_EXPLOITATION_METAUX: 'pxm',
  PERMIS_D_EXPLOITATION_RADIOACTIF: 'pxr',
  PERMIS_D_EXPLOITATION_GRANULATS_MARINS: 'pxw',
} as const satisfies Record<string, (typeof TITRES_TYPES_KEYS)[number]>

export const TitresTypesIds = Object.values(TITRES_TYPES_IDS)

export const titreTypeIdValidator = z.enum(TITRES_TYPES_KEYS)
export type TitreTypeId = z.infer<typeof titreTypeIdValidator>
export const TitresTypes: {
  [key in TitreTypeId]: Definition<key>
} = {
  [TITRES_TYPES_IDS.AUTORISATION_DE_PROSPECTION_CARRIERES]: { id: 'apc', domaineId: 'c', typeId: 'ap' },
  [TITRES_TYPES_IDS.AUTORISATION_DE_PROSPECTION_HYDROCARBURE]: { id: 'aph', domaineId: 'h', typeId: 'ap' },
  [TITRES_TYPES_IDS.AUTORISATION_DE_PROSPECTION_METAUX]: { id: 'apm', domaineId: 'm', typeId: 'ap' },
  [TITRES_TYPES_IDS.AUTORISATION_DE_PROSPECTION_GRANULATS_MARINS]: { id: 'apw', domaineId: 'w', typeId: 'ap' },
  [TITRES_TYPES_IDS.AUTORISATION_DE_RECHERCHE_CARRIERES]: { id: 'arc', domaineId: 'c', typeId: 'ar' },
  [TITRES_TYPES_IDS.AUTORISATION_DE_RECHERCHE_GEOTHERMIE]: { id: 'arg', domaineId: 'g', typeId: 'ar' },
  [TITRES_TYPES_IDS.AUTORISATION_DE_RECHERCHE_METAUX]: { id: 'arm', domaineId: 'm', typeId: 'ar' },
  [TITRES_TYPES_IDS.AUTORISATION_D_EXPLOITATION_METAUX]: { id: 'axm', domaineId: 'm', typeId: 'ax' },
  [TITRES_TYPES_IDS.CONCESSION_FOSSILES]: { id: 'cxf', domaineId: 'f', typeId: 'cx' },
  [TITRES_TYPES_IDS.CONCESSION_GEOTHERMIE]: { id: 'cxg', domaineId: 'g', typeId: 'cx' },
  [TITRES_TYPES_IDS.CONCESSION_HYDROCARBURE]: { id: 'cxh', domaineId: 'h', typeId: 'cx' },
  [TITRES_TYPES_IDS.CONCESSION_METAUX]: { id: 'cxm', domaineId: 'm', typeId: 'cx' },
  [TITRES_TYPES_IDS.CONCESSION_RADIOACTIF]: { id: 'cxr', domaineId: 'r', typeId: 'cx' },
  [TITRES_TYPES_IDS.CONCESSION_SOUTERRAIN]: { id: 'cxs', domaineId: 's', typeId: 'cx' },
  [TITRES_TYPES_IDS.CONCESSION_GRANULATS_MARINS]: { id: 'cxw', domaineId: 'w', typeId: 'cx' },
  [TITRES_TYPES_IDS.INDETERMINE_METAUX]: { id: 'inm', domaineId: 'm', typeId: 'in' },
  [TITRES_TYPES_IDS.INDETERMINE_RADIOACTIF]: { id: 'inr', domaineId: 'r', typeId: 'in' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_CARRIERES_CARRIERES]: { id: 'pcc', domaineId: 'c', typeId: 'pc' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_FOSSILES]: { id: 'prf', domaineId: 'f', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_GEOTHERMIE]: { id: 'prg', domaineId: 'g', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_HYDROCARBURE]: { id: 'prh', domaineId: 'h', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_METAUX]: { id: 'prm', domaineId: 'm', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_RADIOACTIF]: { id: 'prr', domaineId: 'r', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_SOUTERRAIN]: { id: 'prs', domaineId: 's', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_EXCLUSIF_DE_RECHERCHES_GRANULATS_MARINS]: { id: 'prw', domaineId: 'w', typeId: 'pr' },
  [TITRES_TYPES_IDS.PERMIS_D_EXPLOITATION_FOSSILES]: { id: 'pxf', domaineId: 'f', typeId: 'px' },
  [TITRES_TYPES_IDS.PERMIS_D_EXPLOITATION_GEOTHERMIE]: { id: 'pxg', domaineId: 'g', typeId: 'px' },
  [TITRES_TYPES_IDS.PERMIS_D_EXPLOITATION_HYDROCARBURE]: { id: 'pxh', domaineId: 'h', typeId: 'px' },
  [TITRES_TYPES_IDS.PERMIS_D_EXPLOITATION_METAUX]: { id: 'pxm', domaineId: 'm', typeId: 'px' },
  [TITRES_TYPES_IDS.PERMIS_D_EXPLOITATION_RADIOACTIF]: { id: 'pxr', domaineId: 'r', typeId: 'px' },
  [TITRES_TYPES_IDS.PERMIS_D_EXPLOITATION_GRANULATS_MARINS]: { id: 'pxw', domaineId: 'w', typeId: 'px' },
} as const

export const isTitreType = (titreType: string | undefined | null): titreType is TitreTypeId => TitresTypesIds.includes(titreType)

export const toTitreTypeId = (titreTypeType: TitreTypeTypeId, domaineId: DomaineId): TitreTypeId => {
  const titreTypeId = `${titreTypeType}${domaineId}`
  if (!isTitreType(titreTypeId)) {
    throw new Error(`le titre type ${titreTypeId} n'est pas reconnu par Camino`)
  }

  return titreTypeId
}

export const getTitreTypeType = (titreType: TitreTypeId): TitreTypeTypeId => {
  const titreTypeType = titreType.substring(0, 2)
  if (isTitreTypeType(titreTypeType)) {
    return titreTypeType
  } else {
    throw new Error(`le titreType ${titreType} n'a pas de titreTypeType connu, cas impossible`)
  }
}

export const getDomaineId = (titreType: TitreTypeId): DomaineId => {
  const domaineId = titreType.substring(2, 3)
  if (isDomaineId(domaineId)) {
    return domaineId
  } else {
    throw new Error(`le titreType ${titreType} n'a pas de domaineId connu, cas impossible`)
  }
}

export const getTitreTypeTypeByDomaineId = (domaineId: DomaineId): TitreTypeTypeId[] =>
  Object.values(TitresTypes)
    .filter(v => v.domaineId === domaineId)
    .map(({ typeId }) => typeId)
