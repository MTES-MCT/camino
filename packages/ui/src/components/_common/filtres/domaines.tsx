import { Pill } from '@/components/_ui/pill'
import { DomaineId } from 'camino-common/src/static/domaines'

export type Props = {
  element: { id: DomaineId; nom: string }
}

export function FiltresDomaines(props: Props) {
  return (
    <span>
      <Pill color={`bg-domaine-${props.element.id}`} class="mr-xs mono">
        {props.element.id}{' '}
      </Pill>
      <span class="cap-first h6 bold">{props.element.nom}</span>
    </span>
  )
}
