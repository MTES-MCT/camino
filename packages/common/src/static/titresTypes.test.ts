import { DomaineId, domainesIds } from './domaines'
import { TitresTypesIds, isTitreType, toTitreTypeId, getTitreTypeType, getDomaineId, getTitreTypeTypeByDomaineId } from './titresTypes'
import { test, expect } from 'vitest'

test('isTitreType', () => {
  expect(isTitreType(null)).toBe(false)
  expect(isTitreType(undefined)).toBe(false)
  for (const titreTypeId of TitresTypesIds) {
    expect(isTitreType(titreTypeId)).toBe(true)
  }
})

test('toTitreTypeId', () => {
  expect(toTitreTypeId('ap', 'm')).toBe('apm')
  expect(() => toTitreTypeId('ap', 'r')).toThrowErrorMatchingInlineSnapshot(`[Error: le titre type apr n'est pas reconnu par Camino]`)
})

test('getTitreTypeType', () => {
  expect(getTitreTypeType('apm')).toBe('ap')
})

test('getDomaineId', () => {
  expect(getDomaineId('apm')).toBe('m')
})

test.each<DomaineId>(domainesIds)('getTitreTypeTypeByDomaineId %p', domaineId => {
  expect(getTitreTypeTypeByDomaineId(domaineId)).toMatchSnapshot()
})
