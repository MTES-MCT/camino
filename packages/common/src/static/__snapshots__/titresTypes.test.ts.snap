// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTitreTypeTypeByDomaineId "c" 1`] = `
Array [
  "ap",
  "ar",
  "pc",
]
`;

exports[`getTitreTypeTypeByDomaineId "f" 1`] = `
Array [
  "cx",
  "pr",
  "px",
]
`;

exports[`getTitreTypeTypeByDomaineId "g" 1`] = `
Array [
  "ar",
  "cx",
  "pr",
  "px",
]
`;

exports[`getTitreTypeTypeByDomaineId "h" 1`] = `
Array [
  "ap",
  "cx",
  "in",
  "pr",
  "px",
]
`;

exports[`getTitreTypeTypeByDomaineId "i" 1`] = `
Array [
  "cx",
  "in",
  "pr",
  "px",
]
`;

exports[`getTitreTypeTypeByDomaineId "m" 1`] = `
Array [
  "ap",
  "ar",
  "ax",
  "cx",
  "in",
  "pr",
  "px",
]
`;

exports[`getTitreTypeTypeByDomaineId "r" 1`] = `
Array [
  "cx",
  "in",
  "pr",
  "px",
]
`;

exports[`getTitreTypeTypeByDomaineId "s" 1`] = `
Array [
  "cx",
  "pr",
]
`;

exports[`getTitreTypeTypeByDomaineId "w" 1`] = `
Array [
  "ap",
  "cx",
  "pr",
  "px",
]
`;
